### PR TITLE
[MLIR][XLA] HLO to LHLO buffer assignment using live variable analysis

### PR DESCRIPTION
**Pull request by ehsan-nt**
This PR adds blocks signature conversion and buffer allocation using live variable analysis.

### CODE REVIEW

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 1]
```diff
@@ -279,6 +282,60 @@ Type ConvertType(Type t) {
 
 }  // namespace
 
+// Converts the blocks signature and allocate buffers for the block arguments.
+void ConvertBlocksSignature(FuncOp& funcOp,
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: This does more than CovertBlockSignature.

File: tensorflow/compiler/mlir/xla/tests/hlo-legalize-to-lhlo.mlir [hunk 2]
```diff
@@ -11,39 +11,39 @@ func @attrs_copy(%operand: memref<2x2xf32>, %result: memref<2x2xf32>) {
...
+// CHECK-NEXT: ^bb1(%[[ARG2:.*]]: memref<10xf32>, %[[ARG3:.*]]: memref<11x11xf32>): // pred: ^bb0
+// CHECK-NEXT: %[[EXP_RESULT:.*]] = alloc() {temp = true} : memref<10xf32>
+// CHECK-NEXT: dealloc %[[ARG3]] : memref<11x11xf32>
+// CHECK-NEXT: "xla_lhlo.exp"(%[[ARG2]], %[[EXP_RESULT]]) : (memref<10xf32>, memref<10xf32>) -> ()
+// CHECK-NEXT: dealloc %[[ARG2]] : memref<10xf32>
+// CHECK-NEXT: %[[BB2_ARG0:.*]] = alloc() {temp = true} : memref<10xf32>
+// CHECK-NEXT: "xla_lhlo.copy"(%[[EXP_RESULT]], %[[BB2_ARG0]]) : (memref<10xf32>, memref<10xf32>) -> ()
+// CHECK-NEXT: dealloc %[[EXP_RESULT]] : memref<10xf32>
+// CHECK-NEXT: br ^bb2(%[[BB2_ARG0]] : memref<10xf32>)
+// CHECK-NEXT: ^bb2(%[[ARG4:.*]]: memref<10xf32>): // pred: ^bb1
```
Comment (pifon2a) [inline tensorflow/compiler/mlir/xla/tests/hlo-legalize-to-lhlo.mlir]: do you need `// pred...` part here and before?

File: tensorflow/compiler/mlir/xla/tests/hlo-legalize-to-lhlo.mlir [hunk 3]
```diff
@@ -11,39 +11,39 @@ func @attrs_copy(%operand: memref<2x2xf32>, %result: memref<2x2xf32>) {
...
+  br ^bb1(%arg0, %arg1: tensor<10xf32>, tensor<11x11xf32>)
+^bb1(%arg2: tensor<10xf32>, %arg3: tensor<11x11xf32>):
+  %0 = "xla_hlo.exp"(%arg2) : (tensor<10xf32>) -> tensor<10xf32>
+  br ^bb2(%0: tensor<10xf32>)
+^bb2(%arg4: tensor<10xf32>):
+  %2 = "xla_hlo.exp"(%arg4) : (tensor<10xf32>) -> tensor<10xf32>
+  return %2 : tensor<10xf32>
+}
+//      CHECK: (%[[ARG0:.*]]: memref<10xf32>, %[[ARG1:.*]]: memref<11x11xf32>, %[[RESULT:.*]]: memref<10xf32>)
+// CHECK-NEXT: %[[BB1_ARG0:.*]] = alloc() {temp = true} : memref<10xf32>
```
Comment (pifon2a) [inline tensorflow/compiler/mlir/xla/tests/hlo-legalize-to-lhlo.mlir]: is `temp=true` needed ?

File: tensorflow/compiler/mlir/xla/tests/hlo-legalize-to-lhlo.mlir [hunk 4]
```diff
@@ -11,39 +11,39 @@ func @attrs_copy(%operand: memref<2x2xf32>, %result: memref<2x2xf32>) {
...
+// CHECK-NEXT: "xla_lhlo.copy"(%[[ARG0]], %[[BB1_ARG0]]) :
+// CHECK-NEXT: %[[BB1_ARG1:.*]] = alloc() {temp = true} : memref<11x11xf32>
+// CHECK-NEXT: "xla_lhlo.copy"(%[[ARG1]], %[[BB1_ARG1]])
+// CHECK-NEXT: br ^bb1(%[[BB1_ARG0]], %[[BB1_ARG1]] : memref<10xf32>, memref<11x11xf32>)
+// CHECK-NEXT: ^bb1(%[[ARG2:.*]]: memref<10xf32>, %[[ARG3:.*]]: memref<11x11xf32>): // pred: ^bb0
+// CHECK-NEXT: %[[EXP_RESULT:.*]] = alloc() {temp = true} : memref<10xf32>
+// CHECK-NEXT: dealloc %[[ARG3]] : memref<11x11xf32>
```
Comment (pifon2a) [inline tensorflow/compiler/mlir/xla/tests/hlo-legalize-to-lhlo.mlir]: i think i am missing smth: can `arg3` dealloc be done before `exp_result` alloc?

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 5]
```diff
@@ -279,6 +275,44 @@ Type ConvertType(Type t) {
...
+    if (block.getNumSuccessors() > 0) {
+      auto terminatorOp = block.getTerminator();
+      auto loc = terminatorOp->getLoc();
+      rewriter.setInsertionPoint(terminatorOp);
+
+      for (auto operand : terminatorOp->getOperands()) {
+        auto t = operand.getType().dyn_cast<ShapedType>();
+        auto memrefType = MemRefType::get(t.getShape(), t.getElementType());
+        auto alloc = rewriter.create<AllocOp>(loc, memrefType);
+        alloc.setAttr(kTempBufferAttr, rewriter.getBoolAttr(true));
```
Comment (pifon2a) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: @herhut: I did not find any usage of "temp" attribute.

I do not think it is used anywhere, let's get rid of tempbufferattr altogether.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 6]
```diff
@@ -279,6 +275,44 @@ Type ConvertType(Type t) {
...
+        rewriter.create<xla_lhlo::CopyOp>(loc, operand, alloc);
+        terminatorOp->replaceUsesOfWith(operand, alloc);
+      }
+    }
+
+    if (block.isEntryBlock()) continue;
+
+    // All blocks signature must be converted except for the entry block.
+    for (int i = 0; i < block.getNumArguments(); i++) {
```
Comment (pifon2a) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: `for (int i = 0, e = block.getNumArguments(); i < e; ++i)`

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 7]
```diff
@@ -423,6 +446,39 @@ struct RedundantCopiesRemoval : mlir::FunctionPass<RedundantCopiesRemoval> {
...
+        builder.create<DeallocOp>(memref.getLoc(), memref);
+      }
+    }
+  }
+
+  void runOnFunction() override {
+    auto function = getFunction();
+    Liveness liveness(function);
+    for (auto& block : function.getBlocks()) {
+      if (!block.isEntryBlock()) {
```
Comment (pifon2a) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: `if (block.isEntryBlock()) continue;
for ...

**Comment by gbaned** [comment]
@dfki-ehna Can you please address the pifon2a's comments and resolve conflicts? Thanks!

**Pull request closed without merging**